### DISCUSSION TITLE
So why are there not Assert.Less/LessEqual/Greater/GreaterEqual?

### DISCUSSION THREAD

**Discussion by Happypig375** [xunit/xunit#2143] [General]
I see that https://xunit.net/docs/comparisons lists Assert.True as an alternative. However, as with Assert.Equal/NotEqual/InRange, these methods can provide better error messages. This is why NUnit has them. Moreover, I am surprised that nobody has raised this before. Why are these not there?

**Comment by bradwilson** [comment]
Here's the original reasoning why we didn't do this: it doesn't read well.

For example, compare this to:

```csharp
Assert.True(x > 12, "x should be more than 12");
```

vs.

```csharp
Assert.Greater(12, x);
```

This follows the standard pattern of `(expected, actual)` for parameters.

**Reply by tylermontneyacc** [reply]
> This follows the standard pattern of (expected, actual) for parameters.

Which does? I assume you mean "`Assert.True` is closer to the standard pattern than `Assert.Greater`". `Assert.True` is just a condition, there is no expected vs actual, you have to read into it. If anything, `Assert.Greater` is closer as one thinks "I expect 12 to be the greater, but the actual is 13".

I see `Assert.Greater`, I go "x > y". I see `Assert.True`, I go "What's true? Oh, this condition. What's this condition do?". The former is definitely more readable.

**Reply by bradwilson** [reply]
I mean the general pattern for all our assertions that accept two values is `(expected, actual)`. This reads strangely when the English is "I assert that X is greater than 12" translates into `Assert.Greater(12, x)` because that's how all the dual arguments work.

You are correct that `Assert.True` is an outlier in that it one of the much smaller number of assertions whose expected value is in the name of the assertion itself rather than a parameter, but you can also say that `Assert.True(x)` is the same thing as `Assert.Equal(true, x)`. Equality is commutative, but greater/less is not, so while writing `Assert.Equal(x, true)` is the same thing as `Assert.Equal(true, x)` (other than the message being wrong, it's definitely not the case that `Assert.Greater(12, x)` and `Assert.Greater(x, 12)` are the same thing--they're definitely not.

This confusion about arguments is the primary reason why we don't offer Greater or Less. But if it bothers you, we offer our assertion library as source (via NuGet or Git Submodule) and you could easily extend it to add these. We have an example of how to do this here: https://github.com/xunit/samples.xunit/tree/main/v3/AssertExtensions/ExtendingAssert

**Comment by FrankRay78** [comment]
> Assert.True(x > 12, "x should be more than 12");

Having embedded error strings surely isn't best practice?
